File: common/src/main/java/net/shiroha233/roadweaver/config/structure/StructureEntry.java
package net.shiroha233.roadweaver.config.structure;

import net.minecraft.resources.Identifier;

import java.util.Collections;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

/**
 * 单个结构条目，用于 GUI 展示和选择
 */
public final class StructureEntry implements Comparable<StructureEntry> {
    private final Identifier id;
    private final String displayName;
    private final boolean isVanilla;
    private final Set<Identifier> dimensions;

    public StructureEntry(Identifier id, String displayName, boolean isVanilla) {
        this(id, displayName, isVanilla, Set.of());
    }

    public StructureEntry(Identifier id, String displayName, boolean isVanilla, Set<Identifier> dimensions) {
        this.id = id;
        this.displayName = displayName;
        this.isVanilla = isVanilla;
        this.dimensions = dimensions == null ? Set.of() : Collections.unmodifiableSet(new HashSet<>(dimensions));
    }

    public Identifier id() { return id; }
    public String displayName() { return displayName; }
    public boolean isVanilla() { return isVanilla; }
    public Set<Identifier> dimensions() { return dimensions; }
    public String namespace() { return id.getNamespace(); }
    public String path() { return id.getPath(); }

    @Override
    public int compareTo(StructureEntry other) {
        if (this.isVanilla != other.isVanilla) {
            return this.isVanilla ? -1 : 1;
        }
        int nsCompare = this.namespace().compareTo(other.namespace());
        if (nsCompare != 0) return nsCompare;
        return this.path().compareTo(other.path());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof StructureEntry that)) return false;
        return Objects.equals(id, that.id);
    }

    @Override
    public int hashCode() { return Objects.hash(id); }

    @Override
    public String toString() { return id.toString(); }
}

File: common/src/main/kotlin/net/shiroha233/roadweaver/config/structure/StructureEntry.kt
package net.shiroha233.roadweaver.config.structure

import net.minecraft.resources.ResourceLocation
import java.util.Objects

/**
 * 单个结构的条目信息
 *
 * 用于在 GUI 中展示和选择结构
 */
class StructureEntry(
    private val id: ResourceLocation,
    private val displayName: String,
    @get:JvmName("isVanillaFlag")
    val isVanilla: Boolean,
    val dimensions: Set<ResourceLocation> = emptySet()
) : Comparable<StructureEntry> {

    fun id(): ResourceLocation = id

    fun displayName(): String = displayName

    fun isVanilla(): Boolean = isVanilla

    fun dimensions(): Set<ResourceLocation> = dimensions

    /**
     * 获取结构的命名空间
     */
    fun namespace(): String = id.namespace

    /**
     * 获取结构的路径（不含命名空间）
     */
    fun path(): String = id.path

    override fun compareTo(other: StructureEntry): Int {
        // 原版优先，然后按命名空间，最后按路径排序
        if (this.isVanilla() != other.isVanilla()) {
            return if (this.isVanilla()) -1 else 1
        }
        val nsCompare = this.namespace().compareTo(other.namespace())
        if (nsCompare != 0) return nsCompare
        return this.path().compareTo(other.path())
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is StructureEntry) return false
        return Objects.equals(id, other.id)
    }

    override fun hashCode(): Int = Objects.hash(id)

    override fun toString(): String = id.toString()
}
